### ISSUE TITLE
Intermittent log entries showing API connections refused

### ISSUE THREAD

**Issue by ahrenstein** [hjdhjd/homebridge-unifi-protect#129]
**Describe The Problem:**
I have Homebridge connected to UniFi Protect and a video doorbell exposed in HomeKit. It seems to work. The logs for Homebridge are showing API connections refused though.
The CloudKey and Homebridge are on different VLANs but the firewall permits Homebridge the ability to talk to the CloudKey.

**To Reproduce:**
Deploy a CloudKey Gen2+ with UniFi Protect and a UniFi G4 Doorbell.
Deploy Homebridge via Docker on Ubuntu 18.04
Deploy the plugin and connect HomeKit and Protect

**Logs:**

```
[10/6/2020, 5:22:33 PM] [HB Supervisor] OS: Linux 4.15.0-118-generic x64
[10/6/2020, 5:22:33 PM] [HB Supervisor] Node.js v12.18.4 /usr/local/bin/node
[10/6/2020, 5:22:33 PM] [HB Supervisor] Homebridge Path: /usr/local/lib/node_modules/homebridge/bin/homebridge
[10/6/2020, 5:22:33 PM] [HB Supervisor] UI Path: /usr/local/lib/node_modules/homebridge-config-ui-x/dist/bin/standalone.js
[10/6/2020, 5:22:33 PM] [HB Supervisor] Starting Homebridge with extra flags: -I -P /homebridge/node_modules
[10/6/2020, 5:22:33 PM] [HB Supervisor] Started Homebridge v1.2.3 with PID: 356
[10/6/2020, 5:22:33 PM] Loaded config.json with 0 accessories and 2 platforms.
[10/6/2020, 5:22:33 PM] ---
[10/6/2020, 5:22:34 PM] Loaded plugin: homebridge-dummy@0.4.1
[10/6/2020, 5:22:34 PM] Registering accessory 'homebridge-dummy.DummySwitch'
[10/6/2020, 5:22:34 PM] ---
[10/6/2020, 5:22:34 PM] Loaded plugin: homebridge-unifi-protect@3.7.8
[10/6/2020, 5:22:34 PM] Registering platform 'homebridge-unifi-protect.UniFi Protect'
[10/6/2020, 5:22:34 PM] ---
[10/6/2020, 5:22:34 PM] Loaded plugin: homebridge-config-ui-x@4.28.0
[10/6/2020, 5:22:34 PM] Registering platform 'homebridge-config-ui-x.config'
[10/6/2020, 5:22:34 PM] ---
[10/6/2020, 5:22:34 PM] Loading 2 platforms...
[10/6/2020, 5:22:34 PM] [Config] Initializing config platform...
[10/6/2020, 5:22:34 PM] [Config] Running in Service Mode
[10/6/2020, 5:22:34 PM] [UniFi Protect] Initializing UniFi Protect platform...
Setup Payload:
X-HM://[REDACTED]
Enter this code with your HomeKit app on your iOS device to pair with Homebridge:
                       
    ┌────────────┐     
    │ [REDACTED]│     
    └────────────┘     
                       
[10/6/2020, 5:22:34 PM] Homebridge v1.2.3 is running on port 52744.
[10/6/2020, 5:22:34 PM] [UniFi Protect] 10.8.5.5: Controller API API connection refused.
[10/6/2020, 5:22:35 PM] [Config] Homebridge Config UI X v4.28.0 is listening on :: port 8080
[10/6/2020, 5:22:35 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Connected to the Protect controller API (address: 10.8.5.5 mac: [REDACTED]).
[10/6/2020, 5:22:35 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Discovered camera: Front Door [UVC G4 Doorbell] (address: 10.8.6.3 mac: [REDACTED]).
[10/6/2020, 5:22:35 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller refresh interval set to 5 seconds.
[10/6/2020, 5:22:35 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS] Front Door [UVC G4 Doorbell]: Stream quality configured: High.
[10/6/2020, 5:22:35 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS] Front Door [UVC G4 Doorbell]: Discovered doorbell message switch: LEAVE PACKAGE AT DOOR.
[10/6/2020, 5:22:35 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS] Front Door [UVC G4 Doorbell]: Discovered doorbell message switch: DO NOT DISTURB.
[10/6/2020, 5:22:40 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: A doorbell has been detected. Setting the controller refresh interval to 2 seconds.
[10/6/2020, 5:29:32 PM] [UniFi Protect] Invalid login credentials given. Please check your login and password.
[10/6/2020, 5:29:32 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Unable to retrieve NVR configuration information from UniFi Protect. Will retry again later.
[10/6/2020, 5:29:34 PM] [UniFi Protect] 10.8.5.5: Controller API API connection refused.
[10/6/2020, 5:29:35 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Connected to the Protect controller API (address: 10.8.5.5 mac: [REDACTED]).
[10/6/2020, 5:59:35 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/6/2020, 6:02:13 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS] Front Door [UVC G4 Doorbell]: HomeKit video stream request received: 1280x720, 30 fps, 299 kbps.
[10/6/2020, 6:02:26 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS] Front Door [UVC G4 Doorbell]: Stopped video streaming session.
[10/6/2020, 6:29:37 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/6/2020, 6:59:37 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/6/2020, 7:29:39 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
```

**Homebridge Configuration:**

```json
{
    "bridge": {
        "name": "Homebridge",
        "username": "[REDACTED]",
        "port": 52744,
        "pin": "[REDACTED]"
    },
    "accessories": [],
    "platforms": [
        {
            "name": "Config",
            "port": 8080,
            "auth": "form",
            "theme": "dark-mode",
            "tempUnits": "c",
            "lang": "auto",
            "platform": "config"
        },
        {
            "controllers": [
                {
                    "address": "10.8.5.5",
                    "password": "[REDACTED]",
                    "username": "[REDACTED]"
                }
            ],
            "platform": "UniFi Protect",
            "options": [
                "Disable",
                "Enable.MAC.[REDACTED]"
            ]
        }
    ]
}
```

**Screenshots:**
N/A

**Environment:**

* **Homebridge Version:** 1.2.3
* **Node Version:** v12.18.4
* **Homebridge-UniFi-Protect Plugin Version:** v3.7.8
* **Apple Device and iOS / macOS / iPadOS / tvOS Version:** iPhone 11 Pro
* **Operating System and OS Version:** Docker on Ubuntu 18.04
* **UniFi Protect Controller Version:**1.13.4

**Comment by ahrenstein** [comment]
I'm going to add some additional log details as the plugin seems to lose access to the controller randomly.

Additional log details:

```
[10/6/2020, 6:02:13 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS] Front Door [UVC G4 Doorbell]: HomeKit video stream request received: 1280x720, 30 fps, 299 kbps.
[10/6/2020, 6:02:26 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS] Front Door [UVC G4 Doorbell]: Stopped video streaming session.
[10/6/2020, 6:29:37 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/6/2020, 6:59:37 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/6/2020, 7:29:39 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/6/2020, 7:59:41 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/6/2020, 8:29:42 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/6/2020, 8:59:42 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/6/2020, 9:29:43 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/6/2020, 9:59:43 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/6/2020, 10:29:45 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/6/2020, 10:59:47 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/6/2020, 11:29:47 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/6/2020, 11:59:47 PM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/7/2020, 12:29:48 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/7/2020, 12:59:48 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/7/2020, 1:29:49 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/7/2020, 1:59:49 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/7/2020, 2:29:49 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/7/2020, 2:59:50 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/7/2020, 3:29:52 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/7/2020, 3:59:53 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/7/2020, 4:29:54 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/7/2020, 4:59:54 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/7/2020, 5:29:56 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/7/2020, 5:59:57 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/7/2020, 6:29:58 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/7/2020, 6:59:59 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/7/2020, 7:29:59 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/7/2020, 7:59:59 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/7/2020, 8:30:01 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/7/2020, 9:00:01 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/7/2020, 9:30:02 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/7/2020, 9:46:48 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS] Front Door [UVC G4 Doorbell]: HomeKit video stream request received: 1920x1080, 30 fps, 802 kbps.
[10/7/2020, 9:47:06 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS] Front Door [UVC G4 Doorbell]: Stopped video streaming session.
```

```
[10/7/2020, 9:47:44 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS] Front Door [UVC G4 Doorbell]: Stopped video streaming session.
[10/7/2020, 10:00:04 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/7/2020, 10:30:06 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API API connection refused.
[10/7/2020, 11:07:33 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS] Front Door [UVC G4 Doorbell]: HomeKit video stream request received: 1280x720, 30 fps, 299 kbps.
[10/7/2020, 11:07:40 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS] Front Door [UVC G4 Doorbell]: Stopped video streaming session.
[10/7/2020, 11:21:15 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Controller API connection reset.
[10/7/2020, 11:21:15 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Unable to retrieve NVR configuration information from UniFi Protect. Will retry again later.
[10/7/2020, 11:21:17 AM] [UniFi Protect] 10.8.5.5: Controller API API connection refused.
[10/7/2020, 11:21:19 AM] [UniFi Protect] 10.8.5.5: Controller API API connection refused.
[10/7/2020, 11:21:21 AM] [UniFi Protect] 10.8.5.5: Controller API API connection refused.
[10/7/2020, 11:21:23 AM] [UniFi Protect] 10.8.5.5: Controller API API connection refused.
[10/7/2020, 11:21:25 AM] [UniFi Protect] 10.8.5.5: Controller API API connection refused.
[10/7/2020, 11:21:27 AM] [UniFi Protect] 10.8.5.5: Controller API API connection refused.
[10/7/2020, 11:21:29 AM] [UniFi Protect] 10.8.5.5: Controller API API connection refused.
[10/7/2020, 11:21:33 AM] [UniFi Protect] UniFi-Controller [UCK-G2-PLUS]: Connected to the Protect controller API (address: 10.8.5.5 mac: [REDACTED]).
```

**Comment by hjdhjd** [comment]
First, thanks for the detailed report, but this is not a bug in the plugin. The errors are correct. Your UCK is resetting it's Protect instance every so often it seems, or the traffic is being blocked via firewall rules.

The plugin is behaving exactly as it should - when it loses the connection to the controller, it tries to reconnect with an incremental backoff over time.

I can't replicate it in an identical setup on my end. However, I've seen this happen before when a UCKgen2 is nearing the end of it's life. It tends to happen when the drive is about to fail. 😨 

Best of luck.

**Comment by ahrenstein** [comment]
Thanks for the response. This UCKgen2 I've had for about 2 weeks, so that's a little concerning. Do you happen to know where in the UCKgen2 I can find any logs to determine if it's the drive (which is trivial to fix) or a larger issue?

Thanks

**Comment by hjdhjd** [comment]
Poke around in the Ubiquiti forums. I bit the bullet and upgraded to the UDMP / UNVR in my case. :smile:

Another option can be a clean factory reset of the UCKgen2+...if you've ever run beta firmware on the UCK, it can flake out without a full reset.

The usual logs /var/log/messages, etc...are good starting points.

The only quirky thing I find of interest in your setup is that your disconnected happened every 30 minutes like clockwork initially.

Another thing to eliminate...firewalls. Eliminate it entirely from the equation for the time being, just to see what happens. It may aid in troubleshooting.

I love the UCKgen2+, it's an awesome little device. But...and here's the rub...it's definitely underpowered to handle serious use of the drive and handle a lot of traffic (lots of cameras). I had two die on me completely...fried drives and gen2.

Those are the areas I'd look at...can you try eliminating the firewall first?

**Comment by ahrenstein** [comment]
Unfortunately all of the HomeKit stuff is on a different VLAN from the CloudKey, and changing either all HomeKit devices to the UniFi VLAN or rebuilding all UniFi to the HomeKit VLAN is a huge effort. If this is a common CloudKey issue do you recommend moving to the UniFI Protect NVR? (I was considering doing that anyway to support better storage options)

**Comment by hjdhjd** [comment]
UNVR is a vastly superior and more responsive option. More to the point - the UCKgen2+ is based on an older OS for Ubiquiti. Newer devices (UDMP, UNVR, UNVR Pro), all use UniFiOS.

They've said at some point UCKgen2+ will get UniFiOS, but there's no timeline.

Why is that relevant?

UniFiOS has one critical feature that is of great appeal to us...a realtime update API that we can tap into to create true, realtime motion, doorbell, and other notifications. The UCKgen2+ OS lacks the API endpoints to do this unfortunately. As a result, we have to poll the UCKgen2+ for events...when you have a doorbell installed, `homebridge-unifi-protect` will aggressively poll the UCK every two seconds so it doesn't miss a doorbell ring, and it can feel responsive to you, the end user.

TL;DR: buy a newer generation product like a UDMP or a UNVR, if you can. You'll be happier. It's better supported. UNVR in particular (which I use), is terrific. I have 20+ cameras streaming to it and it barely breaks a sweat.

On the firewall note...all you need to do is move the UCK. You don't need to move the cameras. As long as the UCK can talk to the HomeKit VLAN, you're fine. You're just moving a single device.

**Comment by ahrenstein** [comment]
Sounds like the NVR is the way to go. Moving UCK is something I don't want to do since several switches and APs point to it as their controller. I wanted to move to the NVR anyway. Now I'm wondering if the UCK is gonna be an issue as a network controller as well...

I appreciate the info and assistance.

**Comment by hjdhjd** [comment]
UCK is fine as a network controller. It's really the beating the drive is going to take with being a DVR. That said, UDMP is definitely a more robust solution than UCK, in my experience.

Closing this issue.
